### PR TITLE
fix: align auth failure error messages for oauth

### PR DESCRIPTION
**Pull request by cmars**
In the Typescript CLI, when OAuth fails due to an invalid or expired token, throw the same "missing auth" error that one would get if not logged in (no auth token).

The resulting error message will prompt the user to re-authenticate.

This has to be mediated by the legacycli http proxy because the Typescript side of the CLI is not involved in the OAuth flow at all; it is handled transparently by the go-application-framework's networking middleware.

A special HTTP request and response header is used by the legacycli proxy to indicate whether that middleware auth has failed. If it has, the functions which perform HTTP client requests with needle will throw an error that indicates "unauthenticated", resulting in an experience consistent with the legacy token auth flow in the CLI.

Outstanding TODOs on this:

- [x] Improve the header names
- [x] Improve the middleware auth error checking -- a specific error match would be better here 
- [x] Review the approach with the team
- [x] Tests?

---

## Pull Request Submission

Please check the boxes once done.

The pull request must:

- **Reviewer Documentation**
    - [ ] follow [CONTRIBUTING](https://github.com/snyk/cli/blob/main/CONTRIBUTING.md) rules
    - [ ] be accompanied by a detailed description of the changes
    - [ ] contain a risk assessment of the change (Low | Medium | High) with regards to breaking existing functionality. A change e.g. of an underlying language plugin can completely break the functionality for that language, but appearing as only a version change in the dependencies.
    - [ ] highlight breaking API if applicable
    - [ ] contain a link to the automatic tests that cover the updated functionality.
    - [ ] contain testing instructions in case that the reviewer wants to manual verify as well, to add to the manual testing done by the author.
    - [ ] link to the link to the PR for the User-facing documentation
- **User facing Documentation**
    - [ ] update any relevant documentation in gitbook by submitting a gitbook PR, and including the PR link here
    - [ ] ensure that the message of the final single commit is descriptive and prefixed with either `feat:` or `fix:` , others might be used in rare occasions as well, if there is no need to document the changes in the release notes. The changes or fixes should be described in detail in the commit message for the changelog & release notes.
- **Testing**
    - [ ] Changes, removals and additions to functionality must be covered by acceptance / integration tests or smoke tests - either already existing ones, or new ones, created by the author of the PR.

## Pull Request Review

All pull requests must undergo a thorough review process before being merged. 
The review process of the code PR should include code review, testing, and any necessary feedback or revisions. 
Pull request reviews of functionality developed in other teams only review the given documentation and test reports. 

Manual testing will not be performed by the reviewing team, and is the responsibility of the author of the PR.

For Node projects: It’s important to make sure changes in `package.json` are also affecting `package-lock.json` correctly.

****************************If a dependency is not necessary, don’t add it.**************************** 

When adding a new package as a dependency, make sure that the change is absolutely necessary. We would like to refrain from adding new dependencies when possible.
Documentation PRs in gitbook are reviewed by Snyk's content team. They will also advise on the best phrasing and structuring if needed.

## Pull Request Approval

Once a pull request has been reviewed and all necessary revisions have been made, it is approved for merging into 
the main codebase. The merging of the code PR is performed by the code owners, the merging of the documentation PR 
by our content writers.

## What does this PR do?

See top posting.

## How should this be manually tested?

- Build this branch locally.
- Authenticate with OAuth.
- "Damage" the token as follows:
  - Set `expired` in the past
  - Flip some characters in the refresh token, making it invalid.
- Observe that with this binary, all test commands will fail with the same error message as legacy token auth. Other than Code and SBOM, they'll tell you to `snyk auth` (that's a pre-existing issue that will be fixed in a followup).

## What are the relevant tickets?

CLI-392

### CODE REVIEW

File: cliv2/internal/proxy/proxy.go [hunk 1]
```diff
@@ -146,6 +146,7 @@ func (p *WrapperProxy) ProxyInfo() *ProxyInfo {
 func (p *WrapperProxy) replaceVersionHandler(r *http.Request, ctx *goproxy.ProxyCtx) (*http.Request, *http.Response) {
 	err := p.addHeaderFunc(r)
 	if err != nil {
```
Comment (cmars) [inline cliv2/internal/proxy/proxy.go:148]: Once https://github.com/snyk/go-application-framework/pull/213 lands,

```suggestion
	if errors.Is(err, middleware.ErrAuthenticationFailed) {
```

would be more accurate.

File: cliv2/internal/proxy/proxy.go [hunk 2]
```diff
@@ -143,9 +145,28 @@ func (p *WrapperProxy) ProxyInfo() *ProxyInfo {
...
+//
+// The response header is then used by the Typescript CLI to surface an
+// appropriate authentication failure error back to the user.
+//
+// These layers of indirection are necessary because the Typescript CLI is not
+// involved in OAuth authentication at all, but needs to know that an auth
+// failure specifically occurred. HTTP status and error catalog codes aren't
+// adequate for this purpose because there are non-authentication reasons an API
+// request might 401 or 403, such as permissions or entitlements.
+const headerSnykAuthFailed = "x-snyk-auth-failed"
```
Comment (j-luong) [inline cliv2/internal/proxy/proxy.go:163]: question: do we want to stick with the `x-` prefix to denote custom header fields? The convention was [deprecated](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers) AFAIK.

File: cliv2/internal/proxy/proxy.go [hunk 3]
```diff
@@ -143,9 +145,28 @@ func (p *WrapperProxy) ProxyInfo() *ProxyInfo {
 	}
 }
 
+// headerSnykAuthFailed is used to indicate there was a failure to establish
```
Comment (j-luong) [inline cliv2/internal/proxy/proxy.go:148]: praise: appreciate the context given here

File: cliv2/internal/proxy/proxy.go (hunk 2 above)
Comment (cmars) [reply cliv2/internal/proxy/proxy.go:163]: Old habits die hard. Good shout; per [RFC 6648](https://datatracker.ietf.org/doc/html/rfc6648) I think it's reasonable to drop the `x-` prefix here.

**Comment by j-luong** [comment]
LGTM:
Screenshot 2024-07-09 at 17 34 05

**Review by j-luong — APPROVED** [review]

File: cliv2/internal/proxy/proxy.go [hunk 4]
```diff
@@ -177,6 +198,12 @@ func (p *WrapperProxy) Start() error {
 	proxy.Logger = log.New(&pkg_utils.ToZeroLogDebug{Logger: p.DebugLogger}, "", 0)
 	proxy.OnRequest().DoFunc(p.replaceVersionHandler)
 	proxy.OnRequest().HandleConnect(p)
+	proxy.OnResponse().DoFunc(func(resp *http.Response, ctx *goproxy.ProxyCtx) *http.Response {
+		if authFailed := resp.Request.Header.Get(headerSnykAuthFailed); authFailed != "" {
+			resp.Header.Set(headerSnykAuthFailed, authFailed)
```
Comment (PeterSchafer) [inline cliv2/internal/proxy/proxy.go:203]: Question: Why can't we just print the message here? Instead of delegating this to the TS CLI. This way we could get to a consistent error written in the Extensible CLI.

Comment (PeterSchafer) [reply cliv2/internal/proxy/proxy.go:203]: So basically what I'm wondering is, if we can't handle auth errors transparently to the TS CLI, similar to how we handle auth in general.

**Comment by PeterSchafer** [comment]
FYI there is a failing test, which doesn't expect the new http header.

File: cliv2/internal/proxy/proxy.go (hunk 4 above)
Comment (cmars) [reply cliv2/internal/proxy/proxy.go:203]: I worry about inter-process complexities with such an approach, at least while we're still dispatching command execution and console interactions entirely to the TS CLI.

Decomposing console interactions to the Go CLI and treating the TS CLI like a library would mitigate this risk, but that'll take followups beyond the scope of this PR.

If we did error out here in the middleware, I think we'd have to terminate the node subprocess immediately to prevent its output from getting mixed up with that of the parent process. I worry about the assumptions we'd be making about the order of operations scheduled in node's event loop, as well as what's been buffered for output in that process, how it handles signals, and resource cleanup.

Things can get tricky with inter-process coordination. It might be fine in the typical case, but hard to debug and reproduce edge cases which could be timing or platform-dependent.

**Comment by cmars** [comment]
> FYI there is a failing test, which doesn't expect the new http header.

Fixed

**Review by sandor-trombitas — APPROVED** [review]

File: cliv2/internal/proxy/proxy.go (hunk 4 above)
Comment (PeterSchafer) [reply cliv2/internal/proxy/proxy.go:203]: I understand the points and I should probably take one step back from proposing a solution to proposing/aligning on the goal.

My goal is to achieve a consistent behaviour for authentication failures throughout the CLI, no matter if the domain logic is implemented in TS or Golang and in a not too distant future for LS as well.

Some non-functional requirements for a solution would be, Single source of Truth and KISS (Keep it simple stupid).
